enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,891 @@
+
+When these features are identified, we must prioritise adding the feature in Cluster API as these will become blockers for migration.
+Once the feature is implemented in Cluster API, we can unblock the synchronisation for the feature and allow users to migrate.
+
+#### Multiple controllers reconcile the same infrastructure
+
+We aim to prevent multiple controllers (being Machine API and Cluster API) simultaneously reconciling infrastructure as described in the [pausing section][pausing] above.
+
+However, were there to be multiple controllers reconciling the same Machine, there are three cases to consider. Create, Update and Deletion operations.
+
+In a Create operation, if both controllers attempted the create simultaneously, it could lead to an additional host being created and then orphaned within the infrastructure provider.
+Once a controller has created the host, it is expected to populate the provider ID and use this to look up the infrastructure in future reconciles.
+The synchronisation controller would overwrite the non-authoritative provider ID, effectively orphaning the host.
+
+In an Update operation, since the infrastructure is immutable, the updates are only gathering information.
+In this case, the synchronisation controller would overwrite the information, should it differ, in the non-authoritative API.
+This could lead to a hot loop/fight between the two controllers.
+Rate limiting should be used in the controllers to prevent rapid reconciles in this fashion.
+
+In a Delete operation, both controllers would attempt to delete the same infrastructure.
+It is expected that this would not cause issue in most cases as the infrastructure provider should be able to handle multiple requests to delete the same object.
+Where Machine lifecycle hooks are in place on Machines, these should be mirrored across both APIs and should prevent either taking action until they are removed.
+Machine lifecycle hooks are expected to continue to function as expected.
+
+From this, it is imperative that we ensure Create operations are not reconciled by multiple Machine controllers to prevent the potential of leaking resources.
+Testing will be focused on the creation flow.
+
+[pausing]: #pausing-of-cluster-api-controllers
+
+### Drawbacks
+
+#### Mirroring resources may be confusing for end users
+
+We want to provide an easy method for migration for users, however, having multiple representations in OpenShift of the same infrastructure, may be confusing for end users.
+
+They may expect to be able to make changes to either resource and have them reflected.
+For that to be safe we would have to make sure all writes to objects are synchronous to both objects.
+This is technically very challenging and has been dismissed as an alternative.
+
+Documentation will be provided to explain to users the expected behaviours and the expectations to allow them to migrate their infrastructure across.
+
+## Design Details
+
+### Open Questions
+
+N/A
```

Comment:
What is the scope of feature gaps? I think that's something we will have to find out as we implement conversions per platform, it's difficult to answer here.

Reply:
Yeah, I'll sync up with you on this later, but, as we go through and write the code out that converts `providerSpec` to `InfraTemplate`, I'm expecting we will find gaps.

As we find gaps, if the field is in use, and we can't convert, we will have to return an error which will end up on the status in the Synchronized condition.

Every time we write an error case, a Jira story will be created (assuming disciplined engineers) and that will then need to be resolved. Once resolved, the error case can be removed